File: sources/entity/tests/test_northdata.php
<?php
/**
 * Test suite for NorthData authentication and financial data extraction.
 * Run: php tests/test_northdata.php
 */

require_once __DIR__ . '/../php/tools.php';

$config = json_decode(file_get_contents(__DIR__ . '/../php/settings.json'), true);
$tools = new LookupTools($config);

$passed = 0;
$failed = 0;

function test(string $label, $expected, $actual): void
{
    global $passed, $failed;
    if ($expected === $actual) {
        echo "  PASS: {$label}\n";
        $passed++;
    } else {
        echo "  FAIL: {$label}\n    Expected: " . json_encode($expected) . "\n    Got:      " . json_encode($actual) . "\n";
        $failed++;
    }
}

function testContains(string $label, string $needle, string $haystack): void
{
    global $passed, $failed;
    if (str_contains($haystack, $needle)) {
        echo "  PASS: {$label}\n";
        $passed++;
    } else {
        echo "  FAIL: {$label} — expected to contain \"{$needle}\"\n    Got: " . substr($haystack, 0, 300) . "\n";
        $failed++;
    }
}

function testNotContains(string $label, string $needle, string $haystack): void
{
    global $passed, $failed;
    if (!str_contains($haystack, $needle)) {
        echo "  PASS: {$label}\n";
        $passed++;
    } else {
        echo "  FAIL: {$label} — should NOT contain \"{$needle}\"\n";
        $failed++;
    }
}

echo "=== NorthData Tests ===\n\n";

// ── 1. Authentication ──────────────────────────────────────────────────────

echo "1. NorthData Login:\n";
// Use reflection to test private method
$ref = new ReflectionClass($tools);
$method = $ref->getMethod('getNorthdataAuthCookie');
$method->setAccessible(true);
$cookie = $method->invoke($tools);
test('Returns auth cookie', true, $cookie !== null);
test('Cookie is JWT', true, $cookie !== null && substr_count($cookie, '.') === 2);
echo "   Cookie: " . substr($cookie ?? '', 0, 50) . "...\n\n";

// ── 2. Financial extraction from authenticated HTML (ABCA Group) ────────

echo "2. Parse financials — ABCA Systems Group Ltd (auth HTML):\n";
$authHtml = file_get_contents(__DIR__ . '/fixtures/abca_group_auth.html');
test('Auth fixture loaded', true, strlen($authHtml) > 50000);

$parseMethod = $ref->getMethod('parseNorthdataHtml');
$parseMethod->setAccessible(true);
$result = $parseMethod->invoke($tools, $authHtml);

testContains('Has company name', 'Abca Systems Group Ltd', $result);
testContains('Has registry ID', 'Companies House 12500353', $result);
testContains('Has Financials section', '### Financials', $result);
testContains('Has Revenue', 'Revenue', $result);
testContains('Has Revenue value £11M', '£11M', $result);
testContains('Has Earnings', 'Earnings', $result);
testContains('Has Total assets', 'Total assets', $result);
echo "\n   Full result:\n";
echo $result . "\n\n";

// ── 3. Financial extraction from unauthenticated HTML (ABCA Group) ──────

echo "3. Parse financials — ABCA Systems Group Ltd (NO auth HTML):\n";
$noAuthHtml = file_get_contents(__DIR__ . '/fixtures/abca_group_noauth.html');
test('NoAuth fixture loaded', true, strlen($noAuthHtml) > 50000);

$noAuthResult = $parseMethod->invoke($tools, $noAuthHtml);

testContains('Has company name', 'Abca Systems Group Ltd', $noAuthResult);
// Without auth, we should NOT have financials — verify
// (This documents expected behaviour: no auth = no premium data)
echo "   No-auth result:\n";
echo $noAuthResult . "\n\n";

// ── 4. Financial extraction from authenticated HTML (ABCA Ltd) ──────────

echo "4. Parse financials — Abca Systems Ltd (auth HTML):\n";
$ltdHtml = file_get_contents(__DIR__ . '/fixtures/abca_ltd_auth.html');
test('Ltd fixture loaded', true, strlen($ltdHtml) > 50000);

$ltdResult = $parseMethod->invoke($tools, $ltdHtml);

testContains('Has company name', 'Abca Systems Ltd', $ltdResult);
testContains('Has Financials section', '### Financials', $ltdResult);
testContains('Has Revenue', 'Revenue', $ltdResult);
testContains('Has Earnings', 'Earnings', $ltdResult);
testContains('Has Total assets', 'Total assets', $ltdResult);
echo "\n   Full result:\n";
echo $ltdResult . "\n\n";

// ── 5. extractNorthdataFinancials directly ──────────────────────────────

echo "5. extractNorthdataFinancials — isolated test on auth HTML:\n";
$finMethod = $ref->getMethod('extractNorthdataFinancials');
$finMethod->setAccessible(true);

$authFin = $finMethod->invoke($tools, $authHtml);
echo "   Auth financials result:\n{$authFin}\n";
test('Auth financials not empty', true, strlen($authFin) > 0);
if (strlen($authFin) > 0) {
    testContains('Has Revenue row', 'Revenue', $authFin);
    testContains('Has Earnings row', 'Earnings', $authFin);
}

$noAuthFin = $finMethod->invoke($tools, $noAuthHtml);
echo "\n   No-auth financials result:\n{$noAuthFin}\n";
echo "   (Empty is expected — premium data gated)\n\n";

// ── 6. Live search with auth ────────────────────────────────────────────

echo "6. Live searchNorthdata — ABCA Systems Group Ltd:\n";
$liveResult = $tools->searchNorthdata('ABCA Systems Group Ltd');
testContains('Has company name', 'Abca Systems Group Ltd', $liveResult);
testContains('Has Financials section', '### Financials', $liveResult);
testContains('Has Revenue', 'Revenue', $liveResult);
echo "\n   Live result:\n";
echo $liveResult . "\n\n";

// ── Summary ─────────────────────────────────────────────────────────────

echo "=== Results: {$passed} passed, {$failed} failed ===\n";
exit($failed > 0 ? 1 : 0);

File: sources/entity/app/tests/test_northdata.py
"""
Faithful port of ../../tests/test_northdata.php.

NorthData authentication + financial extraction.

PHP method mapping (PHP used ReflectionClass for the privates):
  getNorthdataAuthCookie      -> _get_northdata_auth_cookie()
  parseNorthdataHtml          -> _parse_northdata_html(html)
  extractNorthdataFinancials  -> _extract_northdata_financials(html)
  searchNorthdata             -> search_northdata(name)

Fixture-parse tests (PHP sections 2-5) are PURE — _parse_northdata_html and
_extract_northdata_financials make no network calls — so they run by default.
Login (section 1) and live search (section 6) are gated behind RUN_LIVE=1.

Fixtures live at sources/entity/tests/fixtures/ (shared with the PHP suite).
"""
import os
import pathlib

import pytest

from config import load_config
from tools import LookupTools

_FIXTURES = pathlib.Path(__file__).resolve().parent.parent.parent / "tests" / "fixtures"


@pytest.fixture(scope="module")
def tools():
    return LookupTools(load_config())


@pytest.fixture(scope="module")
def auth_html():
    return (_FIXTURES / "abca_group_auth.html").read_text()


@pytest.fixture(scope="module")
def noauth_html():
    return (_FIXTURES / "abca_group_noauth.html").read_text()


@pytest.fixture(scope="module")
def ltd_html():
    return (_FIXTURES / "abca_ltd_auth.html").read_text()


live = pytest.mark.skipif(not os.environ.get("RUN_LIVE"), reason="live network")


# ══ 1. Authentication (LIVE) ════════════════════════════════════════════════

@pytest.mark.live
@live
def test_northdata_login(tools):
    cookie = tools._get_northdata_auth_cookie()
    assert cookie is not None                        # Returns auth cookie
    assert cookie.count(".") == 2                     # Cookie is JWT


# ══ 2. Parse financials — ABCA Systems Group Ltd (auth HTML) — PURE ═════════

def test_auth_fixture_loaded(auth_html):
    assert len(auth_html) > 50000                     # Auth fixture loaded


def test_parse_auth_html(tools, auth_html):
    result = tools._parse_northdata_html(auth_html)
    assert "Abca Systems Group Ltd" in result         # Has company name
    assert "Companies House 12500353" in result       # Has registry ID
    assert "### Financials" in result                 # Has Financials section
    assert "Revenue" in result                        # Has Revenue
    assert "£11M" in result                           # Has Revenue value £11M
    assert "Earnings" in result                       # Has Earnings
    assert "Total assets" in result                   # Has Total assets


# ══ 3. Parse financials — ABCA Systems Group Ltd (NO auth HTML) — PURE ══════

def test_noauth_fixture_loaded(noauth_html):
    assert len(noauth_html) > 50000                   # NoAuth fixture loaded


def test_parse_noauth_html(tools, noauth_html):
    result = tools._parse_northdata_html(noauth_html)
    assert "Abca Systems Group Ltd" in result         # Has company name
    # Without auth, no premium financials are expected (documents behaviour).


# ══ 4. Parse financials — Abca Systems Ltd (auth HTML) — PURE ═══════════════

def test_ltd_fixture_loaded(ltd_html):
    assert len(ltd_html) > 50000                      # Ltd fixture loaded


def test_parse_ltd_html(tools, ltd_html):
    result = tools._parse_northdata_html(ltd_html)
    assert "Abca Systems Ltd" in result               # Has company name
    assert "### Financials" in result                 # Has Financials section
    assert "Revenue" in result                        # Has Revenue
    assert "Earnings" in result                       # Has Earnings
    assert "Total assets" in result                   # Has Total assets


# ══ 5. extractNorthdataFinancials directly — PURE ═══════════════════════════

def test_extract_financials_auth(tools, auth_html):
    auth_fin = tools._extract_northdata_financials(auth_html)
    assert len(auth_fin) > 0                           # Auth financials not empty
    assert "Revenue" in auth_fin                       # Has Revenue row
    assert "Earnings" in auth_fin                       # Has Earnings row


def test_extract_financials_noauth(tools, noauth_html):
    # Empty is expected — premium data gated. Just ensure it does not raise.
    noauth_fin = tools._extract_northdata_financials(noauth_html)
    assert isinstance(noauth_fin, str)


# ══ 6. Live search with auth (LIVE) ═════════════════════════════════════════

@pytest.mark.live
@live
def test_search_northdata_live(tools):
    result = tools.search_northdata("ABCA Systems Group Ltd")
    assert "Abca Systems Group Ltd" in result          # Has company name
    assert "### Financials" in result                  # Has Financials section
    assert "Revenue" in result                         # Has Revenue
